server/src/main/java/org/elasticsearch/search/crossproject/CrossProjectIndexResolutionValidator.java

```diff
@@ -143,42 +166,85 @@ public static ElasticsearchException validate(
                 // checking if flat expression matched remotely
                 for (String remoteExpression : remoteExpressions) {
                     String[] splitResource = splitQualifiedResource(remoteExpression);
-                    ElasticsearchException exception = checkSingleRemoteExpression(
+                    var projectAlias = splitResource[0];
+                    var resource = splitResource[1];
+
+                    ElasticsearchException remoteException = checkSingleRemoteExpression(
                         remoteResolvedExpressions,
-                        splitResource[0], // projectAlias
-                        splitResource[1], // resource
+                        projectAlias,
+                        resource,
                         remoteExpression,
                         indicesOptions
                     );
-                    if (exception == null) {
+                    if (remoteException == null) {
                         // found flat expression somewhere
                         foundFlat = true;
+                        if (remoteAuthorizationExceptions != null) {
+                            remoteAuthorizationExceptions.clear();
+                            remoteUnauthorizedIndices.clear();
+                        }
                         break;
                     }
-                    if (unauthorizedException == null && exception instanceof ElasticsearchSecurityException securityException) {
+                    if (unauthorizedException == null && remoteException instanceof ElasticsearchSecurityException securityException) {
                         unauthorizedException = securityException;
+                        if (remoteAuthorizationExceptions == null) {
+                            remoteAuthorizationExceptions = new HashMap<>();
+                            remoteUnauthorizedIndices = new HashMap<>();
+                        }
+                        remoteAuthorizationExceptions.putIfAbsent(projectAlias, securityException);
+                        remoteUnauthorizedIndices.computeIfAbsent(projectAlias, k -> new ArrayList<>()).add(resource);
                     }
                 }
                 if (foundFlat) {
                     continue;
                 }
-                // Prefer reporting 403 over 404
                 if (unauthorizedException != null) {
-                    return unauthorizedException;
-                }
-                // Prefer reporting 404 on origin over linked projects
-                if (localException != null) {
+                    if (localAuthorizationException == null) {
+                        localAuthorizationException = unauthorizedException;
+                        localUnauthorizedIndices = new ArrayList<>();
+                    }
+                    localUnauthorizedIndices.add(originalExpression);
```

Comment:
I think this is incorrect because `unauthorizedException` can be populated for a remote project. But this will always report 403 for the origin project. For example, if `logs` is not found on origin and unauthorized on a remote `P1`, this will report 403 for `logs` instead of `P1:logs`.

Reply:
This used to check that there was a local exception, but that seems to have got lost when merging your changes in - I've restored that check